Fix #8456: Convert "throw <string>" to "throw <Error>

abhayraghuwanshi opened the pull request:
## Explanation
Fixes #8456: Replaced
`throw 'Invalid length for list editor: ' + $scope.len;`
with
`throw new Error('Invalid length for list editor: ' + $scope.len);`
everywhere
- The `.toThrow()` is standardised to `toThrowError()`
- The `throw Error` is standardised to `throw new Error/TypeError/DimensionError`
-the `throw 'string'` is standardised to `throw new Error('string')`
- The `throw err(as variable)` is standardised to `throw new Error(err as variable)`

Here is modied file code to test the outcome of newly added linter test
```
Line 101:     }).toThrow('User is not allowed to edit this collection.');
Line 147:     throw 'error';
Line 161:     throw Error('error');
Line 185:      var err = 'error';
Line 186:    throw err;
````
here i my results from terminal
```
/home/abhay/open-source/oppia/core/templates/domain/collection/CollectionRightsObjectFactorySpec.ts --> Line 101: Please use 'toThrowError' instead of 'toThrow'

/home/abhay/open-source/oppia/core/templates/domain/collection/CollectionRightsObjectFactorySpec.ts --> Line 161: Please use 'throw new' instead of 'throw'

/home/abhay/open-source/oppia/core/templates/domain/collection/CollectionRightsObjectFactorySpec.ts --> Line 147: Please use 'throw new Error' instead of 'throw'

/home/abhay/open-source/oppia/core/templates/domain/collection/CollectionRightsObjectFactorySpec.ts --> Line 186: Please use 'throw new Error' instead of 'throw'
```
## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PR CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Commit:
throw string

Commit:
Merge remote-tracking branch 'upstream/develop' into throw-string-error

"Merged with develop"

Commit:
trailing error fixed

Commit:
error solved

kevintab95 commented:
@abhayraghuwanshi, please follow the checklist and edit the PR accordingly. This will help us review your PR in an efficient manner. Thanks!

Commit:
Merge remote-tracking branch 'upstream/develop' into throw-string-error

merged with develop

seanlip commented on core/templates/dev/head/components/forms/schema-based-editors/schema-based-list-editor.directive.ts:
```diff
@@ -225,11 +225,13 @@ angular.module('oppia').directive('schemaBasedListEditor', [
           } else {
             if ($scope.len <= 0) {
-              throw 'Invalid length for list editor: ' + $scope.len;
+              throw new Error( 'Invalid length for list editor: ' +
```
If something in parentheses spans more than one line, please break after the opening parentheses '('.

Ditto elsewhere. Please fix.

abhayraghuwanshi replied:
okay. I will fix it
thanks

seanlip reviewed (changes requested):
Hi @abhayraghuwanshi -- @kevintab95 is currently away for a bit, but I took a quick look and had a general comment that applies in multiple places. Could you PTAL and fix it so that, once he comes back, he can take a pass at the updated version? Thanks.

Commit:
parentheses fixes

Commit:
Merge remote-tracking branch 'upstream/develop' into throw-string-error

merged with develop

seanlip commented:
@kevintab95 PTAL when you get a chance. Thanks!

kevintab95 commented on core/tests/data/third_party/js/third_party.js:
```diff
@@ -35652,7 +35652,9 @@ function GridListDirective($interpolate, $mdConstant, $mdGridLayout, $mdMedia) {
       var colCount = parseInt($mdMedia.getResponsiveAttribute(attrs, 'md-cols'), 10);
       if (isNaN(colCount)) {
-        throw 'md-grid-list: md-cols attribute was not found, or contained a non-numeric value';
+        throw new Error(
```
Need not make changes in third_party.js

abhayraghuwanshi replied:
Thanks
I'll reset it to original

kevintab95 reviewed (changes requested):
Hi @abhayraghuwanshi, looks good in general. Added one comment PTAL. Thanks!

Commit:
fixed the third party

Commit:
Merge remote-tracking branch 'upstream/develop' into throw-string-error

merged with develop

Commit:
exploration error fixed

Commit:
removed ff from describe

seanlip reviewed (approved):
Approving for CODEOWNER files only (core/templates/dev/head/expressions). Thanks @abhayraghuwanshi!

kevintab95 reviewed (approved):
LGTM, thanks @abhayraghuwanshi!

bansalnitish reviewed (approved):
LGTM for codeowner files!

seanlip commented on core/templates/dev/head/pages/exploration-editor-page/services/exploration-property.service.spec.ts:
```diff
@@ -61,7 +61,7 @@ describe('Exploration Property Service', function() {
     expect(function() {
       ExplorationPropertyService.init('initial value');
-    }).toThrow('Exploration property name cannot be null.');
+    }).toThrow(new Error('Exploration property name cannot be null.'));
```
Here and elsewhere, I think you can use toThrowError?

abhayraghuwanshi replied:
i fixed those file in my recent commit
thanks

seanlip reviewed (changes requested):
Hi @abhayraghuwanshi, took a full pass, PTAL. 

Also, just a thought -- we need to prevent this from happening in the future, too. Could you perhaps add a no-throw-literal rule to eslint to prevent this from happening in the future? https://eslint.org/docs/rules/no-throw-literal

Please also make sure to test manually that your new rule works by introducing an erroneous throw statement and seeing if the error is caught by the automated checks, so that authors of PRs who introduce the bad pattern can fix it. Otherwise, we'll end up with dirty code again. Thanks!

Commit:
fixed the to throw to tothrowerror

abhayraghuwanshi commented:
@seanlip in my recent PRs same things are failing on the e2e_topic_and story_editor #8588

kevintab95 commented:
Hi @abhayraghuwanshi, this is due to a flaky test. We're currently trying to figure out a fix for this. Thanks!

abhayraghuwanshi commented:
@kevintab95 Thanks for letting me know.

seanlip commented:
@abhayraghuwanshi please don't forget to respond to the second and third paragraphs in this comment: https://github.com/oppia/oppia/pull/8522#pullrequestreview-356344990. Thanks!

abhayraghuwanshi commented:
@seanlip i will fix that and let you know if i get stuck somewhere
Thanks

abhayraghuwanshi commented:
@seanlip here is the regular expression I made https://regex101.com/r/T85GWZ/14
but it is only matching 1 toThrow(new Error

Commit:
regular expression made

seanlip commented:
@abhayraghuwanshi I don't understand your last comment. If you have a specific question, please clearly state what it is, otherwise we won't be able to help you. Thanks.

Commit:
added

abhayraghuwanshi commented:
@seanlip 
![Screenshot from 2020-02-17 15-45-01](https://user-images.githubusercontent.com/45534131/74685420-836e1880-519c-11ea-801f-edec987420b2.png)
here are the matches of toThrow(new Error via the regular expression I made in https://regex101.com/r/T85GWZ/17 online editor. the file contains 2 
![Screenshot from 2020-02-17 15-48-39](https://user-images.githubusercontent.com/45534131/74685629-1909a800-519d-11ea-8857-7aa14cce7de7.png) the has two matches in the editor
but when I added the 
    {
        'regexp': re.compile(r'\btoThrow\s*\(\n*\s*\bnew Error'),
        'message': 'In tests, please use \'.toThrowError\' instead of',
        'excluded_files': (),
        'excluded_dirs': ()
    }, in the BAD_PATTERNS_JS_AND_TS_REGEXP on the precommitlinter.py file it is only raising one error 
![Screenshot from 2020-02-17 15-54-04](https://user-images.githubusercontent.com/45534131/74685836-bc5abd00-519d-11ea-984d-0b56ed8e842e.png)
it is only considering the case when i used the it like
      }).toThrow(new Error('User is not allowed to unpublish this collection.'));
it is ignoring the case when it is used as
 }).toThrow(
          new Error('User is not allowed to edit this collection.'));

seanlip commented:
Perhaps you could just search for ".toThrow("?

Commit:
merged

Commit:
added the pre commit linter test

Commit:
error fixed

Commit:
fixed

abhayraghuwanshi commented:
@seanlip. i made two cases in the bad expression 
1.if  anyone uses the " throw 'message' "it will  throw an error to use the " throw new Error (message) " instead
2.when they use the toThrow it will raise the error to use toThrowError
Thanks

seanlip reviewed (commented):
Hi @abhayraghuwanshi, I think it looks great. Thanks for standardizing this! Now all you need to do is to fix the lint errors in the CI checks.

abhayraghuwanshi commented:
@seanlip the new linter test i added is causing it. It is failing in 33 files. Should i fix those files too

seanlip commented:
Yes. (Good thing you added the test, right? :) )

abhayraghuwanshi commented:
I'll fix them
Thanks

kevintab95 commented:
Hi @abhayraghuwanshi, just a thought, what happens if we try to do something like `throw <variable>` where `<variable>` could be a string or some other primitive data type assigned to it before the throw statement? Is there any way we can handle this? Thanks.

abhayraghuwanshi commented:
@kevintab95  
In the pre_commit_linter the error is thrown when the specific keywords are matched in the compiled js files. My added tests will fail in case we use **throw**  variable

kevintab95 commented:
Wait, doesn't it currently match `throw '`?

My question was for this scenario:
```ts
var message = 'This is an error message';
throw message;
```

abhayraghuwanshi commented:
yes it is currently matching the  throw '

abhayraghuwanshi commented:
I will fail in that scenario.  it was related to my Pr to change throw string to throw new Error(string

abhayraghuwanshi commented:
@kevintab95 
how to make linter_tests for that case
var message = 'This is an error message';
throw message;
I am not able to figure this one out

kevintab95 commented:
Hm, I'm not too sure but what if we make it such that if we throw an error we only want it to be of the form `throw new Error(...` ie. not allow throwing of variables (even if it means throwing an Error object stored in a variable). Just a thought to try to cover this scenario.

If that is the case, would something like this work?
`\bthrow\s(?!new\sError\()`

Thanks!

abhayraghuwanshi commented:
@kevintab95 
![Screenshot from 2020-02-19 15-55-32](https://user-images.githubusercontent.com/45534131/74875467-4dfa3400-5330-11ea-8d00-ebad6518aa1c.png)
in the above, it is also throwing the error when it used in the string for eg .
  it('should `throw` error when when canUnpublish is false during unpublishing',
    () => {
      var noUnpublishCollectionRightsBackendObject = {
        collection_id: 0,
        can_edit: true,
        can_unpublish: false,
        is_private: false,
        owner_names: ['A']
      };
and it will failed in the case
`   if (arguments.length > 1) {
        throw Error('Second argument for Object.create() is not supported');
      }
      if (o === null) {
        throw Error('Cannot set a null [[Prototype]]');
      }`
and in this too
`  if (this === null) {
        throw new TypeError('this is null or not defined');
      }`

kevintab95 commented:
Can you figure out a way to fix (1) the occurence in the string?
I think (2) and (3) should be straight forward.. You can include them in the regex.

abhayraghuwanshi commented:
@kevintab95 
Do we have to throw error too when it is used as `throw error(msg)`
And when we use it as `throw typeError(msg)`
And when we use it as `throw new typeError (msg)` 
In above cases we have to suggest them to use `throw new error`
And in case he uses `.tothrow(msg)` we throw an error to tell them to use `toThrowError`
And last case when we use `throw e;` we tell to use `throw new Error (e)`

kevintab95 commented:
Hi @abhayraghuwanshi, I took a look at the official Javascript specification [here](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-error-constructor) and it looks like `throw new Error(..` and `throw Error(..` are equivalent. So to standardize let's set it to `new Error(..` in the regex.

Regarding the various other [Error types](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error). Can you do a quick search and find out if we really do use the other types (ie. TypeError, SyntaxError etc...)? If so, we will have to add those to the regex as well.

Hope that helps!

abhayraghuwanshi committed:
all saved

abhayraghuwanshi committed:
mergedd

abhayraghuwanshi commented:
```
   }).toThrow(
      sharedErrorWrapper(
        'This line uses q, so you need to have an earlier line proving ' +
        'that q is true.', 0, 'missing_antecedent', 'logic'));
```
`toThrow(TypeError('eds.getData is not a function'));`
```
.toThrow(
      {
        message: (
          'The name \'we\' is reserved for vocabulary and so cannot ' +
          'be used here.')
      }
    );
```
@kevintab95 how do i handle these . Should i change i change my 
```
  {
        'regexp': re.compile(r'\btoThrow\('),
        'message': 'Please use \'toThrowError\' instead of '
                   '\'toThrow\'',
        'excluded_files': (),
        'excluded_dirs': ()
    },
```

abhayraghuwanshi committed:
all saved

abhayraghuwanshi committed:
saved

abhayraghuwanshi committed:
saved

kevintab95 commented:
I think it is fine to match `toThrow(..` and suggest using `toThrowError(..` instead. Thanks!

abhayraghuwanshi commented:
@kevintab95 my newest commit contains the following
1. Replaced the `throw Error` to `throw new Error`
2. replaced the `toThrow(new Error` to `toThrowError`
3.linter test added  when someone uses the `'throw error/typeError/other_error_type'` it will ask then to use throw new error/typeError instead[as you said to standardise them to use 
`throw new Error/TypeError`
4. linter test for when someone use `throw e;` it will ask them to use `throw new error `instead

abhayraghuwanshi commented:
@kevintab95 could you please take a look
thanks

kevintab95 reviewed (approved):
LGTM thanks @abhayraghuwanshi! Just to check, were you able to test if the linting rules worked for the conditions you described in [this comment](https://github.com/oppia/oppia/pull/8522#issuecomment-589808893)?

Commit:
Merge remote-tracking branch 'upstream/develop' into throw-string-error
merged

abhayraghuwanshi commented:
hii @kevintab95 
Sorry for not replying. My exams are coming so, I am trying to work as much as possible
Referred to the previous comment 
I modified a file to test the newly added linter tests 
the file modified as follows
```
Line 101:     }).toThrow('User is not allowed to edit this collection.');
Line 147:     throw 'error';
Line 161:     throw Error('error');
Line 185:      var err = 'error';
Line 186:    throw err;
````
here i my results from terminal
```
/home/abhay/open-source/oppia/core/templates/domain/collection/CollectionRightsObjectFactorySpec.ts --> Line 101: Please use 'toThrowError' instead of 'toThrow'

/home/abhay/open-source/oppia/core/templates/domain/collection/CollectionRightsObjectFactorySpec.ts --> Line 161: Please use 'throw new' instead of 'throw'

/home/abhay/open-source/oppia/core/templates/domain/collection/CollectionRightsObjectFactorySpec.ts --> Line 147: Please use 'throw new Error' instead of 'throw'

/home/abhay/open-source/oppia/core/templates/domain/collection/CollectionRightsObjectFactorySpec.ts --> Line 186: Please use 'throw new Error' instead of 'throw'
```
Thank you

Commit:
messaged changed

new files changed

abhayraghuwanshi commented:
@kevintab95 should I change the files in the `extensions/interactions/LogicProof/static/js` or add them to excluded directory

kevintab95 commented:
No worries, @abhayraghuwanshi. Good luck with your exams!

Nice work. I think it looks great. Can you resolve the merge conflicts too? Thanks!

kevintab95 commented:
@aks681 @kevinlee12 @vojtechjelinek @nithusha21 codeowner review please. Thanks!

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks!

abhayraghuwanshi commented:
@kevintab95  i added the `extensions/interactions/LogicProof/static/` in the excluded _dir but tests are still failing

abhayraghuwanshi committed:
updated the linter

updated

aks681 reviewed (approved):
Lgtm as codeowner.

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -202,6 +202,32 @@
+        'message': 'Please use \'toThrowError\' instead of '
+                   '\'toThrow\'',
+        'excluded_files': (),
+        'excluded_dirs': ('extensions/interactions/LogicProof/static/js')
```
Why is this file excluded?

abhayraghuwanshi replied:
there are many places in that folder `throw error (as a variable)`. I wasn't sure to I change it or not, so asked here but no one replied so, I excluded the directory. Should I change them?

kevinlee12 replied:
@seanlip second thoughts?

abhayraghuwanshi replied:
and the 

> > ```
> >    }).toThrow(
> >       sharedErrorWrapper(
> >         'This line uses q, so you need to have an earlier line proving ' +
> >         'that q is true.', 0, 'missing_antecedent', 'logic'));
> > ```
> > 
> > 
> > ```
> > .toThrow(
> >       {
> >         message: (
> >           'The name \'we\' is reserved for vocabulary and so cannot ' +
> >           'be used here.')
> >       }
> >     );
> > ```
> 
> @kevinlee12 how do i handle these . My regular expressions are unable to handle these kind of errors
> 
> > ```
> >   {
> >         'regexp': re.compile(r'\btoThrow\('),
> >         'message': 'Please use \'toThrowError\' instead of '
> >                    '\'toThrow\'',
> >         'excluded_files': (),
> >         'excluded_dirs': ()
> > ```

seanlip replied:
Yes, please change it. Having lots of cases is not a sufficient reason to exclude a file/folder from a lint check.

abhayraghuwanshi replied:
@seanlip I was suggested to standardized the `toThrow` to `toThrowError`
according to that should I change the `toThrow(sharedErrorWrapper` to `toThrowError`
and `toThrow({message:(message)}` to `toThrowError`
OR 
should I modified the regex so that it won't catch any error if it used like that`toThrow(sharedErroWrapper`

kevintab95 replied:
Hm, looks like we use Custom errors in those files (and there does not seem to be a way to capture this using regex alone). However, could we maybe enforce that all custom errors should end with Error? @abhayraghuwanshi would this solve the problem for some cases?

abhayraghuwanshi replied:
HI @kevintab95 
and I thought a way to handle the custom error
for example, if we find `toThrowError(something(` using regex we would suggest them to use 
`toThrowError(somethingCustomError` instead

if we change the `node_modules/@types/jasmine/ts3.1/index.d.ts`
```
toThrow(expected?: any): boolean;
toThrowError(message?: string | RegExp): boolean;
``` 
here the `toThrow` is expecting `type:any` but `toThrowError` we can only throw regex or string. and changing the type to any will solve the problem.

So, Do we need to standardized the `toThow` to `toThrowError`  because we cannot handle the objects in this case.

kevintab95 replied:
Er.. I don't think we should be making changes to `node_modules/@types/jasmine/ts3.1/index.d.ts`. Can you give an example of what would fail for the `toThrowError(..` stuff with your current changes? Thanks!

abhayraghuwanshi replied:
when we use` toThrowError(sharedWrapperError` in the case of `toThrow(sharedWrapperError` gives the following error
the shared wrapper function is returning the object
````
    var sharedErrorWrapper = function(message, line, code, category) {
        return {
          message: message,
          line: line,
          code: code,
          category: category
        };
      };
````
````
No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ message: any; line: any; code: any; category: any; }' is not assignable to parameter of type 'new (...args: any[]) => Error'.
      Type '{ message: any; line: any; code: any; category: any; }' provides no match for the signature 'new (...args: any[]): Error'.ts(2769)
index.d.ts(591, 9): The last overload is declared here.
````

kevintab95 replied:
Oh, I see. Thanks @abhayraghuwanshi for pointing that out! I just noticed that there is a TODO in [shared.ts](https://github.com/oppia/oppia/blob/develop/extensions/interactions/LogicProof/static/js/shared.ts#L48) that says we must use Error objects instead of custom objects. In that case, I think it is fine to handle this separately (ie. once the TODO stuff is done, we should enable the linter to lint these files). @seanlip, @kevinlee12  what are your thoughts?

seanlip replied:
Please fix the TODO as part of this PR. (Note that the TODO seems to be referring to the code above it, not the code below it.)

abhayraghuwanshi replied:
@seanlip I'll fix it after 6th, as my exams are going on 
thanks

abhayraghuwanshi replied:
@seanlip  Is this the thing we are talking in the Todo 
https://stackoverflow.com/questions/783818/how-do-i-create-a-custom-error-in-javascript
thanks

seanlip replied:
Yes, something along those lines, but read all the SO answers carefully to get a sense for which approach is the best to adopt.

seanlip replied:
Also, make sure to test the interaction manually after your changes to ensure that everything is still working correctly and that answers can be submitted and evaluated.

abhayraghuwanshi replied:
hii @seanlip I have tried to do the Todo but I am not sure how to proceed further( as the todo says to make the custom error prototype of an Error object) 
````
  var UserError = function(code, parameters) {
    this.name = 'UserError';
    this.code = code;
    this.parameters = parameters || {};
  };
  UserError.prototype = Error.prototype;
  UserError.prototype.name = 'UserError';

  // These errors already have messages which are user-readable.
  var PreRenderedUserError = function(messages, code) {
    this.name = 'PreRenderedUserError';
    this.messages = messages;
    this.code = code;
  };
  PreRenderedUserError.prototype = Error.prototype;
  PreRenderedUserError.prototype.name = 'PreRenderedUserError';
````
should they return the string like an Error.
Could you please clearify it a little more
Thanks

seanlip replied:
Hi @abhayraghuwanshi, thank you for following up on this!

I had a chat with the author of the logic proof interaction for what to do here. We decided that we are going to look into deprecating it, since it is not space-efficient, and we are running into maintenance issues such as the one you are describing here.

Given this decision, I think it is fine to exclude the file for now. Please add a clear comment above the line in the pre-commit linter file to explain why we are doing this, so that other developers know that it is a very special case and that they should not add more things there as well.

Sorry for the back-and-forth on this, but I do want to clarify that the technical information you provided was necessary in order to make this decision! I hope this resolves the conversation, but let me know if you have more questions. Thanks!

abhayraghuwanshi replied:
![Screenshot from 2020-03-11 11-54-32](https://user-images.githubusercontent.com/45534131/76436987-478b3500-638f-11ea-9007-d8d00697b568.png)
@seanlip  there is one more problem with the linter in was told to standardized the 
`throw err;` to `throw new Error(err);` as seen above when the err is an object the error will be thrown like the `Uncaught Error[object object]` something like above 
what should be do when this type of senario arrises

seanlip replied:
If we are ignoring this file, then does this case occur (where the error is not a string)?  If so, where does it occur?

abhayraghuwanshi replied:
https://github.com/oppia/oppia/pull/8522/files#diff-f212e0f6c92196b87df08e0e8e460b1d
https://github.com/oppia/oppia/pull/8522/files#diff-b6e084d7c98d0c8637b0bc5999ef2624
https://github.com/oppia/oppia/pull/8522/files#diff-d8160714a691a346dc4594acdab35777
https://github.com/oppia/oppia/pull/8522/files#diff-fbb9a6ab7eab26c37730e2617360ddc6
these are the list of changes I did in the where the throw err is an object, although I have excluded the logicProof  from the linter but maybe this thing create problem in future 
![Screenshot from 2020-03-12 01-33-04](https://user-images.githubusercontent.com/45534131/76491072-282def80-6403-11ea-9421-e4f23c3d3101.png)

seanlip replied:
Ah, OK. If it is straightforward to change these to strings, and nothing else in the codebase uses the error raised from here and treats it as an object, then it is fine to change these to strings (like all the others). Does that solve the problem?

abhayraghuwanshi replied:
@seanlip i guess this will solve the problem .
If i have any doubts again I'll let you know
Thanks

abhayraghuwanshi replied:
@seanlip i have updated them like 
````
     var additionalInfo = (
          'name: ' + err.name +
          '\nmessage: ' + err.message +
          '\nstack: ' + err.stack
        );
        throw new Error(additionalInfo);
````
PTAL
thanks

kevinlee12 reviewed (approved):
lgtm, though just a question.

Commit:
added

Commit:
Merge remote-tracking branch 'upstream/develop' into throw-string-error
merged

abhayraghuwanshi commented:
> ```
>    }).toThrow(
>       sharedErrorWrapper(
>         'This line uses q, so you need to have an earlier line proving ' +
>         'that q is true.', 0, 'missing_antecedent', 'logic'));
> ```
> 
> ```
> .toThrow(
>       {
>         message: (
>           'The name \'we\' is reserved for vocabulary and so cannot ' +
>           'be used here.')
>       }
>     );
> ```
> 
@kevinlee12  how do i handle these . My regular expressions are unable to handle these kind of errors
> 
> ```
>   {
>         'regexp': re.compile(r'\btoThrow\('),
>         'message': 'Please use \'toThrowError\' instead of '
>                    '\'toThrow\'',
>         'excluded_files': (),
>         'excluded_dirs': ()

nithusha21 reviewed (approved):
LGTM as a codeowner. Thanks

seanlip commented:
Sure thing :) Good luck with your exams!

Commit:
save

abhayraghuwanshi committed:
merged with develop

abhayraghuwanshi committed:
Merge remote-tracking branch 'upstream/develop' into throw-string-error
Merged with develop

abhayraghuwanshi committed:
changed some files

abhayraghuwanshi committed:
minor linting error and yarn removed

yarn removed from commit

abhayraghuwanshi committed:
the error thrown object is changed to string

abhayraghuwanshi committed:
yarn and linter test updated

seanlip commented on core/templates/components/state-directives/answer-group-editor/answer-group-editor.directive.ts:
```diff
@@ -358,7 +358,12 @@ angular.module('oppia').directive('answerGroupEditor', [
               );
               e.message += additionalInfo;
-              throw e;
+              var errorStr = (
```
Actually, for this one, maybe get rid of the try/catch block and do:

```
if (!INTERACTION_SPECS.hasOwnProperty(interactionId)) {
  throw Error(...);
}

return INTERACTION_SPECS[interactionId].is_trainable;
```

abhayraghuwanshi replied:
changed the block of code
thanks

seanlip commented on core/templates/domain/story/story-update.service.ts:
```diff
@@ -57,7 +57,12 @@ angular.module('oppia').factory('StoryUpdateService', [
+          '\nmessage: ' + err.message +
+          '\nstack: ' + err.stack
+        );
+        throw new Error(additionalInfo);
```
Hm. The problem with doing this is that it obscures the stack, and it also prints the stack in the message which is quite unwieldy if the message ever gets shown to the user.

So, I think you'll need to stick with `throw err` here. I think your lint check could just check specifically for the whole line being 'throw X;' and, if so, look back through the previous lines one at a time until you get to a "} catch (X) {" line. If that line exists, it's all good. If you encounter another throw before that, though, or if you reach the beginning of the file, then something is wrong and we have a lint error.

What do you think?

abhayraghuwanshi replied:
@kevinlee12 I will make an expression and let u know
thanks

abhayraghuwanshi replied:
@kevinlee12 I am making a regex which will match `throw error;` first, then it looks back until it finds the `catch(error) {`.  
````
catch(error) {
  code code code;
  code code code
  code code code;
  throw error;
}
````
To achieve this I have to match the whole block of code written below
````
catch(error) {
  code code code;
  code code code
  code code code;
````
I couldn't figure out the way to match this whole block of code.
Can you help me, please
Thanks

kevinlee12 replied:
What have you tried?

abhayraghuwanshi replied:
First, I tried to match this block first. 
 ````
catch(error) {
  code code code;
  code code code
  code code code
````
https://regex101.com/r/38C4ev/1
then I will replace   `(?<!catch)\bthrow\s\w+;`  `catch`  from this expression with the regex matching the block of code above

kevinlee12 replied:
Try this: `catch[(]\w+[)]\s+[{]\n\s+([a-zA-Z0-9_\s;]+)}` you're missing whitespace captures.

abhayraghuwanshi replied:
Thanks. I will let u know if I get stuck again

seanlip commented on core/templates/expressions/expression-syntax-tree.service.spec.ts:
```diff
@@ -42,7 +42,8 @@ describe('Expression syntax tree service', () => {
     it('should throw if environment is not found', () => {
-      expect(() => ExpressionSyntaxTreeService.lookupEnvs('', [])).toThrow();
+      expect(() => ExpressionSyntaxTreeService.lookupEnvs('', []))
+        .toThrowError();
```
No message?

abhayraghuwanshi replied:
Added the message

seanlip commented on core/templates/filters/convert-unicode-with-params-to-html.filter.spec.ts:
```diff
@@ -47,7 +47,7 @@ describe('HTML to text', function() {
           return $filter('convertUnicodeWithParamsToHtml')(s);
         };
-        expect(fn).toThrow();
+        expect(fn).toThrowError();
```
No message? Add messages where possible, to make the assertion stricter.

abhayraghuwanshi replied:
Added the message

seanlip commented on core/templates/pages/collection-editor-page/services/collection-editor-state.service.spec.ts:
```diff
@@ -409,7 +409,7 @@ describe('Collection editor state service', function() {
       expect(function() {
         CollectionEditorStateService.saveCollection('Commit message');
-      }).toThrow();
+      }).toThrowError();
```
Ditto.

abhayraghuwanshi replied:
added the thrown message

seanlip commented on core/templates/pages/exploration-editor-page/services/user-exploration-permissions.service.spec.ts:
```diff
@@ -74,6 +74,6 @@ describe('User Exploration Permissions Service', function() {
     ueps.getPermissionsAsync();
 
-    expect($httpBackend.flush).toThrow();
+    expect($httpBackend.flush).toThrowError();
```
Ditto, here and elsewhere.

abhayraghuwanshi replied:
Added the message

seanlip commented on core/templates/pages/exploration-player-page/layout-directives/progress-nav.directive.ts:
```diff
@@ -98,7 +98,12 @@ angular.module('oppia').directive('progressNav', [
                 '\ninterationId: ' + $scope.interactionId);
               e.message += additionalInfo;
-              throw e;
+              var errorStr = (
```
Let's change this -- see above.

abhayraghuwanshi replied:
Changed it back to the original

seanlip commented on core/templates/pages/exploration-player-page/services/answer-classification.service.spec.ts:
```diff
@@ -233,7 +233,7 @@ describe('Answer classification service with string classifier disabled',
       expect(function() {
         acs.getMatchingClassificationResult(stateName, state.interaction, 0);
-      }).toThrow();
+      }).toThrowError();
```
Ditto here and below

seanlip commented on core/templates/services/assets-backend-api.service.ts:
```diff
@@ -110,7 +110,12 @@ angular.module('oppia').factory('AssetsBackendApiService', [
               );
               e.message += additionalInfo;
-              throw e;
+              var errorStr = (
```
See above

abhayraghuwanshi replied:
Changed it back to throw e;

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -202,6 +202,47 @@
+        'message': 'Please use \'toThrowError\' instead of '
+                   '\'toThrow\'',
+        'excluded_files': (
+            # The toThrow error is standardised toThrowError and currently
```
Can you write this so that it is in proper English? I don't understand the first sentence, and I can't tell what the "you" in the second-last line refers to.

abhayraghuwanshi replied:
updated the comment in the excluded file

seanlip reviewed (changes requested):
Thanks, @abhayraghuwanshi. I have taken a full pass, PTAL!

abhayraghuwanshi commented:
@seanlip 
Sorry for the delay
I am currently working on a proposal now, as soon as I am done I will continue this issue
Thanks

abhayraghuwanshi committed:
merge conflict resolved

abhayraghuwanshi commented:
Hii @seanlip 
This is the code block from the assets-backend-api.service.ts
````
        if (exception.name === 'TypeError' && window.BlobBuilder) {
            try {
              var blobBuilder = new BlobBuilder();
              blobBuilder.append(data);
              assetBlob = blobBuilder.getBlob(assetType.concat('/*'));
            } catch (e) {
              var additionalInfo = (
                '\nBlobBuilder construction error debug logs:' +
                '\nAsset type: ' + assetType +
                '\nData: ' + data
              );
              e.message += additionalInfo;
              throw e;
            }
          } else {
            var additionalInfo = (
              '\nBlob construction error debug logs:' +
              '\nAsset type: ' + assetType +
              '\nData: ' + data
            );
            exception.message += additionalInfo;
            throw exception;
          }
        }
````
we can handle the first case `throw e;` in the above code block by matching `throw e;` and then looking back and if `catch(e) {` is  found then ignoring. Error won't appear in that case in the linter test 
But, how will we handle the else block `throw exception` condition

kevintab95 commented:
Hi @abhayraghuwanshi, I think you can delete the try-catch you mentioned in your [comment](https://github.com/oppia/oppia/pull/8522#issuecomment-606104944).

Actually you can replace the if block with the following code:
```ts
  var blobBuilder = new BlobBuilder();
  blobBuilder.append(data);
  assetBlob = blobBuilder.getBlob(assetType.concat('/*'));
```

Looks like the logs were originally added to investigate #6162. But that error hasn't occured in a while.. I think it was fixed by adding a second argument to the Blob constructor in #7784. So it should be safe to delete the try catch. Let me know if you have any questions.

abhayraghuwanshi commented:
@kevintab95 i will change that
thank you

abhayraghuwanshi commented:
Hi @kevintab95 
I was writing the error message for the core/templates/expressions/expression-syntax-tree.service.spec.ts
the function in the service file
````
  const lookupEnvs = (name: string, envs: object[]) => {
      for (const env of envs) {
        if (env.hasOwnProperty(name)) {
          return env[name];
        }
      }

      throw new ExprUndefinedVarError(name, envs);
    };
````
the custom error
````
    const ExpressionError = function() {};
    ExpressionError.prototype = new Error();
    ExpressionError.prototype.constructor = ExpressionError;

    var ExprUndefinedVarError = function(varname, envs) {
      this.varname = varname;
      this.envs = envs;
    };
    ExprUndefinedVarError.prototype = new ExpressionError();
    ExprUndefinedVarError.prototype.constructor = ExprUndefinedVarError;
    ExprUndefinedVarError.prototype.name = 'ExprUndefinedVarError';
    ExprUndefinedVarError.prototype.toString = function() {
      return this.name + ': ' + this.varname + ' not found in ' + this.envs;
    };
````
for this case 
````
    it('should throw if environment is not found', () => {
      expect(() => ExpressionSyntaxTreeService.lookupEnvs('test', [{}]))
        .toThrowError('ExprUndefinedVarError: test not found in [{}]');
    })
````
the error thrown is empty string 
![Screenshot from 2020-03-31 03-44-12](https://user-images.githubusercontent.com/45534131/78000255-02a94d00-7302-11ea-8575-db2603740ede.png)
should I change the custom error object, so that the message thrown will be like  `'ExprUndefinedVarError: test not found in [{}]'` ?

abhayraghuwanshi commented:
@kevintab95 
`core/templates/filters/convert-unicode-with-params-to-html.filter.spec.ts`
I am writing the message in the `toThrowError()` for the case 
````
  var invalidUnicodeStrings = [
    '{}',
    '}}abc{{',
    '\\{{a}}',
    '{{a\\}}',
    '{{a}\\}'
  ];

  it('should detect invalid unicode strings', angular.mock.inject(
    function($filter) {
      invalidUnicodeStrings.forEach(function(s) {
        var fn = function() {
          return $filter('convertUnicodeWithParamsToHtml')(s);
        };
        expect(fn).toThrowError('Invalid unicode-string-with-parameters: '+ s);
      });
    }));
````
Error from the terminal 
````
Error: Expected function to throw an exception with message 'Invalid unicode-string-with-parameters: {}', but it threw an exception with message 'Invalid unicode-string-with-parameters: false'.
````
Is this the expected outcome ?

abhayraghuwanshi commented:
![Screenshot from 2020-04-01 02-31-44](https://user-images.githubusercontent.com/45534131/78106371-42833980-73c1-11ea-8317-5d2ccf44ec3e.png)
@seanlip @kevintab95 
We can handle the object error in this way. Is it correct?

kevintab95 commented:
Hi @abhayraghuwanshi, for your first comment. I think you will need to update the custom error definition. I suspect the reason may be that the `message` variable is not set inside `ExprUndefinedVarError`. See [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/message), it says that the Error.prototype.message will be an empty string by default (explains the failure!). 

Regarding your second comment, have you tried tracing back through the code to figure out what is expected? If not, I would recommend that you do that to figure out what is the expected output for the function.

abhayraghuwanshi commented:
I have trace back the error and fix it soon.
Regarding the object error can we use
 throw new Error(
      JSON.stringify(object variable )):

On Thu, Apr 2, 2020, 2:44 AM Kevin Thomas <notifications@github.com> wrote:

> Hi @abhayraghuwanshi <https://github.com/abhayraghuwanshi>, for your
> first comment. I think you will need to update the custom error definition.
> I suspect the reason may be that the message variable is not set inside
> ExprUndefinedVarError. See this
> <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/message>,
> it says that the Error.prototype.message will be an empty string by default
> (explains the failure!).
>
> Regrading your second comment, have you tried tracing back through the
> code to figure out what is expected? If not, I would recommend that you do
> that to figure out what is the expected output for the function.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/8522#issuecomment-607492865>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AK3MXM6GQY2BPJFF67DQP7LRKOVCTANCNFSM4KNZYZOA>
> .
>

abhayraghuwanshi committed:
updated the throw message

abhayraghuwanshi committed:
merge with develop

abhayraghuwanshi committed:
linting error

abhayraghuwanshi committed:
lint error

kevintab95 commented:
Hi @abhayraghuwanshi, please resolve the merge conflicts.

Also regarding the `JSON.stringify` stuff, can you give me some context? Where do you want to use it and why?

U8NWXD reviewed (approved):
codeowner files lgtm

abhayraghuwanshi commented:
@kevintab95 
for example 
````
let errorObject = {
  'Name': "ExpressionCustomError",
  'Message': 'Error Message'
};
````
when we use it as `throw errorObject;` the error will be
 `{Name: "ExpressionCustomError", Message: "Error Message"}`

we want to handle the error object by the `throw new Error(errorObject)` 
when we use the error object variable in it,, it will give output as the `[Object Object]`

but when we use it as the `throw new Error(JSON.stringify(errorObject))`it will throw an error like this `Uncaught Error: {"Name":"ExpressionCustomError","Message":"Error Message"}`

we can use it in the files containing the custom errors
extensions/interactions/LogicProof/static/js/tools/teacher2.ts
extensions/interactions/LogicProof/static/js/shared.ts

kevintab95 commented:
Hm, I think one way to go about this would be to try to see if the behaviour would remain the same with your change. You could try reproducing some of the errors you mentioned in the files (using the Logic proof interaction UI) and then try the same with your changes, and check if they behave the same way. If they do, I think you can propose this change in your PR, if not you will need to find some other fix.

abhayraghuwanshi commented:
@kevintab95 
okay I will try to regenerate it and let u know
thanks

abhayraghuwanshi committed:
Fixed the ci linter error

abhayraghuwanshi committed:
merged

kevintab95 commented:
Hi @abhayraghuwanshi, can you try this regex to capture all throws except ones inside a catch block?
`(?s)(?:(?<=catch).*?throw(?=\s.*;))|(throw\s\w;)`

I've made the first group a non-capturing group so it should not be reported. If that doesn't work maybe we could look into tweaking the script to report only captured groups. Hope this helps!

abhayraghuwanshi committed:
regex addded

abhayraghuwanshi committed:
added the parenthesis

abhayraghuwanshi committed:
saved

abhayraghuwanshi committed:
Merged with develop

abhayraghuwanshi committed:
merge conflict resolved

abhayraghuwanshi commented:
@kevintab95 PTAL
Thanks

abhayraghuwanshi commented:
@Hudda PTAL
Thanks

kevintab95 commented:
Per discussion with @abhayraghuwanshi, this PR will be closed and will be taken over by me. Thanks!
